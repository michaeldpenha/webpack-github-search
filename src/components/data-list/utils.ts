import CommentTupple from 'components/tupple/comment-tupple';
import IssueTupple from 'components/tupple/issue-tupple';
import RepoTupple from 'components/tupple/repo-tupple';
import { ITupple, TuppleType } from 'src/interface';

// export const tuppleSelector = (type: Typename): React.FC<ITupple> => {
//   return {
//     IssueComment: CommentTupple,
//     Issue: IssueTupple,
//     Repository: RepoTupple,
//   }[type];
// };

export const tuppleSelector = (type: TuppleType): React.FC<ITupple> => {
  return {
    comment: CommentTupple,
    issue: IssueTupple,
    repo: RepoTupple,
  }[type];
};
