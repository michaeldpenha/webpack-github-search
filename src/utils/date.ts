import formatDistanceToNow from 'date-fns/formatDistanceToNow';

/**
 * This function will return formatted date string
 * @param createdAt {string}
 * @returns {string} string with formatted date
 */
export const formatDistance = (createdAt: string): string =>
  !isNaN(Date.parse(createdAt))
    ? formatDistanceToNow(new Date(createdAt), { addSuffix: true })
    : '';
