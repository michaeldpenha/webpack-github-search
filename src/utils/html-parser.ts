import parse from 'html-react-parser';

/**
 * This function will parse Html string and will return HTML elements using html-react-parser
 * @param val
 * @returns {element | string}
 */
export const parseHTML = (val: string): string | JSX.Element | JSX.Element[] => parse(val);
