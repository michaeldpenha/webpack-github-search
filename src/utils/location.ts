import { History, Location } from 'history';

/**
 * This function will update the url params without refresh
 * @param location location object
 * @param value which needs to be updated
 * @param history history object
 * @param key param key
 * @param append whether to append or to delete
 * @return {void}
 */

export const updaeUrlQueryParams = (
  location: Location,
  value: string,
  history: History,
  key: string,
  append?: boolean,
) => {
  const params: URLSearchParams = new URLSearchParams(location.search);
  params.delete(key);
  if (append) {
    params.append(key, value);
  }
  history.replace({
    search: params.toString(),
  });
};
