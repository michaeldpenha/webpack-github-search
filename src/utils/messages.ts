import { MESSAGES } from 'src/messages';

export const formatMessageText = (key: string, values?: string[]): string => {
  const localeText: string = MESSAGES[key];
  if (!values?.length) {
    return localeText;
  }

  return values.reduce(
    (result: string, value: string, index: number) => result.replace(`{${index}}`, value),
    localeText,
  );
};
