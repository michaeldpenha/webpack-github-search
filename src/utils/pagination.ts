import { IssueComment, SearchResultItemEdge, PageInfo } from 'types/schema';

interface IData {
  edges: Array<IssueComment | SearchResultItemEdge>;
  pageInfo: PageInfo;
  totalCount?: number;
}
/**
 * This function is basically merging edges object for pagination
 * @param currentData
 * @param newData
 * @returns {IData}Merged edges data along with pageInfo
 */
export const mergeData = (currentData: IData, newData: IData): IData => {
  const result: IData = {
    pageInfo: newData.pageInfo,
    edges: currentData ? [...currentData.edges, ...newData.edges] : newData.edges,
    totalCount: newData.totalCount,
  };

  return result;
};
