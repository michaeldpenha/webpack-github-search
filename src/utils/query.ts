import { DEFAULT_ISSUE_SEARCH_QUERY } from 'src/constants';

/**
 * Funciton will return query string which will be used in search query
 * @param params URL search query object
 * @returns {string} query string
 */
export const prepareSearchQuery = (params: URLSearchParams): string => {
  const queryParams = {
    type: DEFAULT_ISSUE_SEARCH_QUERY.type,
    language: params.get('langauge') || DEFAULT_ISSUE_SEARCH_QUERY.language,
    state: params.get('state') || DEFAULT_ISSUE_SEARCH_QUERY.state,
  };

  if (params.get('search')) {
    queryParams['in'] = `title,body ${params.get('search')}`;
  }

  if (params.get('repo')) {
    queryParams['repo'] = params.get('repo');
  }

  const result = Object.keys(queryParams).reduce((prev: string, key: string) => {
    return `${prev} ${key}:${queryParams[key]}`;
  }, '');
  return result;
};
