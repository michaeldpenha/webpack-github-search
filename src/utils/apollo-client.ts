import {
  ApolloClient,
  ApolloLink,
  HttpLink,
  InMemoryCache,
  NormalizedCacheObject,
} from '@apollo/client';

import { config } from '../../config';
import { mergeData } from './pagination';

const { githubGraphQlEndPoint: uri } = config;

const authMiddleware = (): ApolloLink =>
  new ApolloLink((operation, forward) => {
    operation.setContext({
      headers: {
        authorization: `Bearer ${process.env.GITHUB_TOKEN}`,
      },
    });
    return forward(operation);
  });

const cache: InMemoryCache = new InMemoryCache({
  possibleTypes: {
    SearchResultItem: ['Issue', 'Repository'],
  },
  typePolicies: {
    Query: {
      fields: {
        search: {
          keyArgs: ['query'],
          merge(existing, incoming) {
            return mergeData(existing, incoming);
          },
          read(existing) {
            return existing;
          },
        },
      },
    },
    Issue: {
      keyFields: ['id', 'number'],
      fields: {
        comments: {
          keyArgs: [],
          merge(existing, incoming) {
            return mergeData(existing, incoming);
          },
          read(existing) {
            return existing;
          },
        },
      },
    },
    Repository: {
      keyFields: ['name', 'id'],
    },
    IssueComment: {
      keyFields: ['id'],
    },
    User: {
      keyFields: ['login'],
    },
  },
});

const httpLink: HttpLink = new HttpLink({
  uri,
});

// const persistedHttpLink: ApolloLink = createPersistedQueryLink({
//   useGETForHashedQueries: true,
//   sha256,
// }).concat(httpLink);

export const apolloClient: ApolloClient<NormalizedCacheObject> = new ApolloClient({
  cache,
  link: authMiddleware().concat(httpLink),
});
