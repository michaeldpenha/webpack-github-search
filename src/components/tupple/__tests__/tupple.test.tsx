import React from 'react';
import renderEl, { fireEvent } from 'src/testUtils/tests-render';
import { formatDistance } from 'utils/date';
import { parseHTML } from 'utils/html-parser';
import IssueTupple from '../issue-tupple';
import RepoTupple from '../repo-tupple';

const setup = ({ data, Component }) => {
  const onTuppleClick = jest.fn();
  const el = renderEl(<Component data={data} onTuppleClick={onTuppleClick} />);
  return {
    ...el,
    onTuppleClick,
  };
};

describe('Tupple component', () => {
  it('should render repo component', () => {
    const { getByTestId, onTuppleClick } = setup({
      data: {
        id: 'MDEwOlJlcG9zaXRvcnkyMTM1NDI4ODc=',
        name: 'Jenkins_Upgradev3',
        nameWithOwner: 'anshulc55/Jenkins_Upgradev3',
        __typename: 'Repository',
        owner: {
          avatarUrl:
            'https://avatars.githubusercontent.com/u/15612383?u=2de2d038453a1769caffc028a724dcfca29e891f&v=4',
        },
      },
      Component: RepoTupple,
    });

    const el = getByTestId('repo-tupple-MDEwOlJlcG9zaXRvcnkyMTM1NDI4ODc=');

    fireEvent.click(el);
    expect(onTuppleClick).toHaveBeenCalled();

    const imgEl = getByTestId('repo-tupple-image-MDEwOlJlcG9zaXRvcnkyMTM1NDI4ODc=');
    expect(imgEl).toHaveAttribute(
      'src',
      'https://avatars.githubusercontent.com/u/15612383?u=2de2d038453a1769caffc028a724dcfca29e891f&v=4',
    );

    expect(el).toMatchSnapshot();
  });

  it('should render issue compoennt for open Issue', () => {
    const { getByTestId, onTuppleClick } = setup({
      data: {
        id: 'MDU6SXNzdWU4OTIwMTUyNDI=',
        bodyHTML: '<p>DISCORD</p>',
        title: 're',
        author: {
          avatarUrl: 'https://avatars.githubusercontent.com/u/58448036?v=4',
          login: 'NeonGamerBot-QK',
        },
        closed: false,
        closedAt: null,
        createdAt: '2021-05-14T15:25:13Z',
        body: 'DISCORD',
      },
      Component: IssueTupple,
    });

    const el = getByTestId('issue-tupple-MDU6SXNzdWU4OTIwMTUyNDI=');
    const headerEl = getByTestId('issue-tupple-header-MDU6SXNzdWU4OTIwMTUyNDI=');
    const articleEl = getByTestId('issue-tupple-article-MDU6SXNzdWU4OTIwMTUyNDI=');
    const footerEl = getByTestId('issue-tupple-footer-MDU6SXNzdWU4OTIwMTUyNDI=');

    fireEvent.click(el);
    expect(onTuppleClick).toHaveBeenCalled();
    expect(headerEl).toHaveTextContent('re');
    expect(articleEl).toHaveTextContent('DISCORD');
    expect(footerEl).toHaveTextContent(formatDistance('2021-05-14T15:25:13Z'));
  });
});
