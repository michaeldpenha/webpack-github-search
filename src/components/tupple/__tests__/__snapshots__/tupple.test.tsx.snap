// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Tupple component should render repo component 1`] = `
<div
  class="styles__RepoTuppleContainer-sc-1oca8og-2 dSRkdy"
  data-testid="repo-tupple-MDEwOlJlcG9zaXRvcnkyMTM1NDI4ODc="
>
  <p>
    <img
      alt="avatar"
      data-testid="repo-tupple-image-MDEwOlJlcG9zaXRvcnkyMTM1NDI4ODc="
      src="https://avatars.githubusercontent.com/u/15612383?u=2de2d038453a1769caffc028a724dcfca29e891f&v=4"
    />
    anshulc55/Jenkins_Upgradev3
  </p>
</div>
`;
